[Feature] Add basic functionality to the `mention_input_bar` component

ZhangHanDong opened the pull request:
Add basic functionality to the `mention_input_bar` component.   For issue #278 

The `mention_input_bar` component requires these modifications to the CommandTextInput component of Makepad to work :[Makepad PR #651](https://github.com/makepad/makepad/pull/651)

![2](https://github.com/user-attachments/assets/122dcf1d-016b-49b2-847e-c2517c43eab4)
![3](https://github.com/user-attachments/assets/49fca7d8-8962-485a-9eb2-df6e71824d5c)


## TODO for the future:

1. Add a header to the user list to display the current number of users in the room.
2. Implement scrolling functionality for the user list.
3. Enable sorting for the user list to show currently online users.
4. Optimize performance and add a loading animation for the user list.

ZhangHanDong commented:
The current error in CI with cargo check is because the dependent CommandTextInput component has not been updated to the latest version.
This PR needs to be merged. [Makepad PR #651](https://github.com/makepad/makepad/pull/651)

alanpoon commented:
![Screenshot 2025-02-10 at 3 19 16 PM](https://github.com/user-attachments/assets/7c260bff-f2fc-489d-8a8e-dc6c7d9e27dc)

1. Mentionable List should not appear when only the "@" is being typed, it should appear only after "@ and some match characters"
2. There should be a header "Users" at the top of the mention-able List
3. Mentionable List seems to contain current user and does not contain one remaining room user.
4. 
![mention2](https://github.com/user-attachments/assets/530e9f75-7ebe-45ae-a843-49777e104d70)
The typing textbox needs to also be able to show avatar, after selecting the mentioned user.
5. 
![mention3](https://github.com/user-attachments/assets/cd167bf2-775e-4184-8c96-8fc4976f7a30)
The typing textbox is not able to send the correct format for the mentioning the user. After mentioning the user, the message should show a hyperlink for the mentioned user. 

6. 
https://github.com/user-attachments/assets/9f43e74e-7f0d-4ed6-a69c-322396f5a2af
The highlighted background by the mouse should be same color as default highlighted mentionable user/

7. In the mentionable List, the user_id should have a "@" at the front.

ZhangHanDong commented:
> ![Screenshot 2025-02-10 at 3 19 16 PM](https://private-user-images.githubusercontent.com/8205468/411446688-7c260bff-f2fc-489d-8a8e-dc6c7d9e27dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyNTcyMjgsIm5iZiI6MTczOTI1NjkyOCwicGF0aCI6Ii84MjA1NDY4LzQxMTQ0NjY4OC03YzI2MGJmZi1mMmZjLTQ4OWQtOGE4ZS1kYzZjN2Q5ZTI3ZGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTFUMDY1NTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjU1YjhlYTRiNWFlZTJmOGYwNDNkNDM5NWViMDA1ZDQzYThkMjY3NDViNjVmODlmMmIyZjM0NzU5N2FhOTk3OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BhxHF2rd492e6g6YWQ2cxvTGxctEOLiQlc781S4tzXI)
> 
> 1. Mentionable List should not appear when only the "@" is being typed, it should appear only after "@ and some match characters"
> 2. There should be a header "Users" at the top of the mention-able List
> 3. Mentionable List seems to contain current user and does not contain one remaining room user.
> 
> ![mention2](https://private-user-images.githubusercontent.com/8205468/411448322-530e9f75-7ebe-45ae-a843-49777e104d70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyNTcyMjgsIm5iZiI6MTczOTI1NjkyOCwicGF0aCI6Ii84MjA1NDY4LzQxMTQ0ODMyMi01MzBlOWY3NS03ZWJlLTQ1YWUtYTg0My00OTc3N2UxMDRkNzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTFUMDY1NTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDhiMTFkZTA4ZGQzZDRkNzIyNjZlNzhkM2M3NmUyZDU1MmQxNGZhNDhjNGExNjcxNmQxZTg3MjczMDRjMDE4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.9ISVQTgKpE9cRdHw7qeL354ESafaxOa3Ur1YyZQDbA0) The typing textbox needs to also be able to show avatar, after selecting the mentioned user. 5. ![mention3](https://private-user-images.githubusercontent.com/8205468/411448758-cd167bf2-775e-4184-8c96-8fc4976f7a30.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyNTcyMjgsIm5iZiI6MTczOTI1NjkyOCwicGF0aCI6Ii84MjA1NDY4LzQxMTQ0ODc1OC1jZDE2N2JmMi03NzVlLTQxODQtOGM5Ni04ZmM0OTc2ZjdhMzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTFUMDY1NTI4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGQyZTBhZDAwOWZmMDExYzc1YTFmNzU5MzMyZGJkMDgwODQxMWQyNjY4M2NhZjc2YWFkZjMzYTFkMTRlNmVmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zJZiny1TRhnvWFplCTYiwuANKyDT8QudXi-UaOkShGw) The typing textbox is not able to send the correct format for the mentioning the user. After mentioning the user, the message should show a hyperlink for the mentioned user.
> 
> 
> https://github.com/user-attachments/assets/9f43e74e-7f0d-4ed6-a69c-322396f5a2af The highlighted background by the mouse should be same color as default highlighted mentionable user/
> 
> 7. In the mentionable List, the user_id should have a "@" at the front.

I don’t quite understand the reason for the first point. 

I will continue to address the suggestions 4/5/6/7. 

The others can be considered as features to be added in the future. This is just the initial version.

kevinaboos commented:
> The current error in CI with cargo check is because the dependent CommandTextInput component has not been updated to the latest version. This PR needs to be merged. [Makepad PR #651](https://github.com/makepad/makepad/pull/651)

the Makepad PR has been merged in now.

ZhangHanDong commented:
<img width="836" alt="Screenshot 2025-02-24 at 18 30 57" src="https://github.com/user-attachments/assets/42b2b149-09fd-4873-b06b-86acda760c65" />

Add a header and other modifications to the popup user list. But it depends on the new changes of the CommandTextInput component. [PR 663](https://github.com/makepad/makepad/pull/663).

ZhangHanDong commented:
Currently, getting room members will cache a list locally. Is there any room for improvement in this solution? @kevinaboos

kevinaboos commented on src/shared/mention_input_bar.rs:
```diff
@@ -0,0 +1,637 @@
+    /// List of available room members for mentions
+    #[rust]
+    room_members: Vec<RoomMember>,
+    /// Current input text content
+    #[rust]
+    current_input: String,
```
why do we need to duplicate the input text here? It's already stored in the `TextInput`, right?

kevinaboos replied:
looks like `current_input` isn't being used at all.

kevinaboos commented on src/shared/mention_input_bar.rs:
```diff
@@ -0,0 +1,637 @@
+    /// * `room_id` - The Matrix room identifier
+    pub fn set_room_id(&mut self, room_id: OwnedRoomId) {
+        self.room_id = Some(room_id.clone());
+        submit_async_request(MatrixRequest::FetchRoomMembers { room_id });
```
this isn't what we want to call here. `FetchRoomMembers` is a synchronization operation that pulls down the latest list of room members such that events in the timeline have the proper user profile associated with them.

It doesn't have any effect on the list of room members returned by the client, especially since we've already fetched the room members previously when showing the room.

kevinaboos commented on src/shared/mention_input_bar.rs:
```diff
@@ -0,0 +1,637 @@
+        message_input.apply_over(
+            cx,
+            live! {
+                trigger: "@",
```
why would we override the trigger string here?  we'd want to preserve the `#[live]` value set in the DSL block...

kevinaboos replied:
actually, we shouldn't override any of these live values.

kevinaboos commented on src/shared/mod.rs:
```diff
@@ -11,8 +11,10 @@ pub mod text_or_image;
 pub mod typing_animation;
 pub mod popup_list;
 pub mod verification_badge;
+pub mod mention_input_bar;
```
see my other comment: move this out of `src/shared/`

kevinaboos commented on src/sliding_sync.rs:
```diff
@@ -478,23 +480,51 @@ async fn async_worker(
             }
 
             MatrixRequest::FetchRoomMembers { room_id } => {
+                // let (timeline, sender) = {
+                //     let all_room_info = ALL_ROOM_INFO.lock().unwrap();
+                //     let Some(room_info) = all_room_info.get(&room_id) else {
+                //         log!("BUG: room info not found for fetch members request {room_id}");
+                //         continue;
+                //     };
+
+                //     (room_info.timeline.clone(), room_info.timeline_update_sender.clone())
+                // };
+
+                // // Spawn a new async task that will make the actual fetch request.
+                // let _fetch_task = Handle::current().spawn(async move {
+                //     log!("Sending fetch room members request for room {room_id}...");
+                //     timeline.fetch_members().await;
+                //     log!("Completed fetch room members request for room {room_id}.");
+                //     sender.send(TimelineUpdate::RoomMembersFetched).unwrap();
+                //     SignalToUI::set_ui_signal();
+                // });
```
oh i see, you replaced the functionality of `FetchRoomMembers`. Please don't do this, the "fetch members" functionality is required for the timeline to work properly.

Instead, you should add a new `MatrixRequest` variant that calls `Room::members()` and sends the `Vec<RoomMember>` list to the timeline.
The RoomScreen/Timeline widget can cache that `Vec<RoomMember>` list in a field called `latest_room_members` or something like that.

kevinaboos replied:
And then, if you want to rename `MatrixRequest::FetchRoomMembers` (because it might be confusing with your new GetRoomMembers request), we can call it `SyncRoomMemberList` or something like that. That'll help disambiguate between two types of requests:
1. a request to synchronize the internal list of room members
2. a request to return the actual list of joined room members

kevinaboos commented on src/sliding_sync.rs:
```diff
@@ -478,23 +480,51 @@ async fn async_worker(
-                    sender.send(TimelineUpdate::RoomMembersFetched).unwrap();
-                    SignalToUI::set_ui_signal();
+                    let room = timeline.room();
+                    match room.members(RoomMemberships::JOIN).await {
```
this request/function should offer more functionality:
* It should provide an option that first calls `Room::members_no_sync()` in order to provide an initial quick response that doesn't need to contact the homeserver.
* It should then optionally call `Room::members()` to get the latest full list from the homeserver, which provides the (slower) full response.
* It should allow passing in a `RoomMemberships` object instead of hard-coding the `JOIN`-ed members only.

So, the request needs more parameters:
1. the room ID
2. The `RoomMemberships` bitflag
3. a boolean for whether to call `Room::members_no_sync()`
4. a boolean for whether to call `Room::members()`

kevinaboos reviewed (changes requested):
Thanks, great start on this complex feature!

First, there are quite a few places where you're using `chars`/characters, e.g., in `find_mention_trigger_position()`. We can't use characters, we _always_ must use unicode graphemes instead.

Second, I don't think we need a cache for the list of members in a room -- that's not something that is used frequently, so we shouldn't really cache it. The room user list will be cached internally by the client's EventCache, so we don't really want to cache it separately. 
We can just store the latest list of `Vec<RoomMember>`s in the `MentionInputBar`'s `room_members` field (and optionally, also in the `TimelineUiState` struct if needed), we don't need a full cache.

kevinaboos commented:
I recently added support for an EditingPane that allows the user to edit messages. This EditingPane widget contains its own instance of the `TextInput` widget, which is completely separate from the message input_bar that you've modified here in this PR.

We need to support mentioning users in _both_ the existing `input_bar`'s `TextInput` at the bottom of the `RoomScreen`, _and_ in the new `edit_text_input` `TextInput` within the EditingPane. 

Can you refactor this PR to provide a `MentionableTextInput` widget that we can instantiate in multiple places? Then we can try it out in both the RoomScreen's input_bar _and_ in the EditingPane's TextInput.

ZhangHanDong commented:
I have extracted the MentionableTextInput component and added the @ user functionality to the EditPane component.

To share the Members in the MentionableTextInput component across multiple components, I added the RoomMemberManager module, which provides a sub/pub mode, allowing multiple components dependent on the MentionableTextInput component to share member data without copying. The components currently sharing the MentionableTextInput component are RoomInputBar and EditPane.

However, I have encountered a **bug**: when opening multiple windows (two room windows, one above the other), the `room_id` of the two windows gets refreshed to the same value.


Additionally, this pull request also depends on the merge of the makepad `CommandTextInput` [pull request #663](https://github.com/makepad/makepad/pull/663).

Finally, to unify everyone’s Rust fmt format, I added a rustfmt.toml, so that running cargo fmt locally will no longer result in a large number of unrelated files being modified.

kevinaboos commented on src/home/editing_pane.rs:
```diff
@@ -409,74 +494,117 @@ impl EditingPane {
-            .as_ref()
-            .map(|info| info.event_tl_item.clone())
+        let inner = self.borrow()?;
+        if !inner.visible {
+            return None;
+        }
```
why is this needed? 

If we do need this, it would be better to handle it by creating a `fn hide()` which clears the inner `info` field.

ZhangHanDong replied:
I changed it to clear the info when the animation state is hidden.

kevinaboos commented on src/home/room_screen.rs:
```diff
@@ -2641,6 +2631,13 @@ pub enum TimelineUpdate {
     /// though the success or failure of the request is not yet known until the client
     /// requests the member info via a timeline event's `sender_profile()` method.
     RoomMembersFetched,
```
since we changed the name of the request to use the word "Sync", we should also use that here.

```suggestion
    RoomMembersSynced,
```

kevinaboos replied:
(and update the docs accordingly)

kevinaboos commented on src/home/room_screen.rs:
```diff
@@ -2641,6 +2631,13 @@ pub enum TimelineUpdate {
     RoomMembersFetched,
+    /// A notice that the room's full member list has been fetched from the server,
+    /// includes a complete list of room members that can be shared across components.
+    /// This is different from RoomMembersFetched which only indicates members were fetched
```
see above comment
```suggestion
    /// This is different from RoomMembersSynced which only indicates members were fetched
```

kevinaboos commented on src/room/room_input_bar.rs:
```diff
@@ -0,0 +1,244 @@
+        width: Fill,
+        height: Fit
+        flow: Right
+        align: {y: 0.5}
```
it's important for this to be bottom-aligned, as I said in my comment here (which you removed; please add it back in). https://github.com/project-robius/robrix/pull/372/files#diff-024cc17fae2704ce65467a3578e635bd1541bdcacc90de230b6f759b08bab906L831

Does that not work with the new mentionable text input stuff?

ZhangHanDong replied:
ok. I forgot about this place.

kevinaboos commented on src/sliding_sync.rs:
```diff
@@ -522,14 +530,49 @@ async fn async_worker(
                 });
             }
 
+            MatrixRequest::GetRoomMembers { room_id, memberships, use_cache, from_server } => {
```
you can simplify this by combining `use_cache` and `from_server` into one boolean.

Check out how i do this in the `GetUserProfile` request variant, I just use a single `local_only` boolean.

kevinaboos commented on src/home/editing_pane.rs:
```diff
@@ -158,21 +179,73 @@ struct EditingPaneInfo {
+
+/// Implement `RoomMemberSubscriber` trait, receive member update notifications
+impl RoomMemberSubscriber for EditingPaneSubscriber {
+    fn on_room_members_updated(
+        &mut self, cx: &mut Cx, room_id: &OwnedRoomId,
+        members: Arc<Vec<matrix_sdk::room::RoomMember>>,
+    ) {
```
this formatting style is weird. It should either be on one line (if it can fit), like this:
```rust
fn short_func(&self, arg1: S, arg2: B) -> R {
    ...
}
```
or if it's long, each parameter should be on its own line, like so:
```rust
fn on_room_members_updated(
    &mut self,
    cx: &mut Cx,
    room_id: &OwnedRoomId,
    members: Arc<Vec<matrix_sdk::room::RoomMember>>,
) {
    ....
}
```

please configure rustfmt to respect this styling

kevinaboos commented on src/home/editing_pane.rs:
```diff
@@ -409,74 +494,117 @@ impl EditingPane {
+        log!("Creating room member subscription for EditingPane, ID: {:?}", self.widget_uid());
+
+        // Create new subscriber
+        let subscriber =
+            Arc::new(
+                Mutex::new(
+                    EditingPaneSubscriber { widget_uid: self.widget_uid(), current_room_id: Some(room_id.clone())}
+                )
+            );
```
this formatting is particularly weird. It should be something like:
```rust
let subscriber = Arc::new(
    Mutex::new(
        EditingPaneSubscriber {
            widget_uid: self.widget_uid(),
            current_room_id: Some(room_id.clone()),
        }
    )
);
```

or a condensed version: 
```rust
let subscriber = Arc::new(Mutex::new(EditingPaneSubscriber {
    widget_uid: self.widget_uid(),
    current_room_id: Some(room_id.clone()),
}));
```
The point is that we want the closing parenthesis `)` to line up with the first `let` keyword, which makes it easier to read.

kevinaboos commented on src/home/room_screen.rs:
```diff
@@ -1134,7 +1110,10 @@ impl Widget for RoomScreen {
 
+
             // Handle the send message button being clicked or Cmd/Ctrl + Return being pressed.
+            let message_input = self.room_input_bar(id!(input_bar)).text_input(id!(text_input));
```
this needs to be moved higher up in the function. It should replace the let statement in Line 1021, otherwise the prior usage of `message_input` will not work.

So, replace the following line earlier in this function:
```rust
let message_input = self.text_input(id!(message_input));
```
with what you have here. And then you can remove this line that you added here.

kevinaboos commented on src/utils.rs:
```diff
@@ -420,6 +421,104 @@ pub fn get_or_fetch_event_sender(
+/// Safely replaces text between byte indices with a new string,
+/// ensuring proper grapheme boundaries are respected
+pub fn safe_replace_by_byte_indices(text: &str, start_byte: usize, end_byte: usize, replacement: &str) -> String {
+    let text_graphemes: Vec<&str> = text.graphemes(true).collect();
+
+    let start_grapheme_idx = byte_index_to_grapheme_index(text, start_byte);
+    let end_grapheme_idx = byte_index_to_grapheme_index(text, end_byte);
+
+    let before = text_graphemes[..start_grapheme_idx].join("");
+    let after = text_graphemes[end_grapheme_idx..].join("");
+
+    format!("{before}{replacement}{after}")
+}
+
+/// Safe version of String::contains that works with graphemes
+pub fn contains_grapheme(text: &str, pattern: &str) -> bool {
+    // For single grapheme patterns
+    if pattern.graphemes(true).count() == 1 {
+        text.graphemes(true).any(|g| g == pattern)
+    } else {
+        // For multi-grapheme patterns, we can use the default contains
+        // as it will work correctly for complete strings
+        text.contains(pattern)
+    }
+}
+
+/// Builds a mapping array from graphemes to byte positions in the string
+pub fn build_grapheme_byte_positions(text: &str) -> Vec<usize> {
+    let mut positions = Vec::with_capacity(text.graphemes(true).count() + 1);
+    let mut byte_pos = 0;
+
+    positions.push(0);
+
+    for g in text.graphemes(true) {
+        byte_pos += g.len();
+        positions.push(byte_pos);
+    }
+
+    positions
+}
```
are these functions used anywhere? I don't see them being called. If they're not used, please remove them.

ZhangHanDong replied:
These are methods for handling glyphs, in case they are needed in some places in the future. However, I will remove the unused ones for now.

kevinaboos reviewed (changes requested):
Thanks, overall I do think the PR looks good. Is it intended to be merged into `main` in the current state? There are tons of log statements and some things look a bit unfinished -- were you leaving those in until after my final approval, or should i remove myself or let you remove them now before final approval?

My main comment is that I don't think we need such a complicated solution for subscribing to a changed list of room members, since we already have so many mechanisms for inter-thread/task communication (actions, existing queues, channels), .... but we can always simplify that later, so let's not worry about it now. Plus, the code quality of your implementation is very good, so I'm inclined to leave it in for now since we may have future needs for inter-widget synchronization that actually justify that complex pub-sub implementation.

I appreciate the rustfmt configuration, but ideally we would have left those changes to be a separate PR. There are just so many unrelated formatting changes in this PR now that it makes it very hard for me to review the actual functionality changes in isolation from trivial formatting changes. 

Can you undo a lot of the unrelated and unnecessary formatting changes? 

Anyway, since this PR does introduce formatting config, my hand was forced so I did end up going through the entire docs of rustfmt. I have some thoughts and left comments in certain places in the code; also, here are some config options I'd like to add:
```
inline_attribute_width: 50 (or higher, we'll see)
imports_granularity: Crate
group_imports: StdExternalCrate

fn_single_line: true
format_generated_files: false

blank_lines_upper_bound: 3
hex_literal_case: Upper
reorder_modules: false
spaces_around_ranges: true
use_try_shorthand: true
```
Note: please don't apply rustfmt to all files in this PR -- let's do that later.

ZhangHanDong commented on src/home/editing_pane.rs:
```diff
@@ -358,16 +448,14 @@ impl Widget for EditingPane {
-
     fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
-        if self.info.is_none() {
+        if !self.animator_in_state(cx, id!(panel.show)) || self.info.is_none() {
```
removed it.

ZhangHanDong commented:
I have made the corrections according to your review comments. @kevinaboos 

The `rustfmt.toml` content has also been added with the configurations you wanted to include. I did not run cargo fmt for this commit. Let’s keep the `rustfmt.toml` for future use.

kevinaboos reviewed (approved):
Thanks! Looks good now, appreciate all the effort that went into this!
